ls1intum/Artemis
# `Programming exercises`: Fix Jenkins builds by removing all non-printable characters in C exercises

### Checklist
#### General


#### Changes affecting Programming Exercises

### Motivation and Context


Instructors reported multiple invalid submissions in some C (FACT) exercises about printing specific character sequences. It has emerged that the problem is related to special Unicode characters that can be included in the feedbacks, which then make the resulting `tests-results.xml` invalid.  

Specifically in this case, the latter was about `Line Tabulation (U+000B)` and `Form Feed (Ff) (U+000C)` characters. 

### Description

In the Jenkins C templates, some Unicode characters are already removed, but more seem to lead to the bug described above. Using `sed -i 's/[^[:print:]]//g' test-reports/tests-results.xml` in the template, would remove all such 'non-printable' characters, ensuring the creation of valid XML test-results files.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 C Programming Exercise (already created in TS2)

1. Log in to Artemis using "Test User 11"
2. Navigate to the C exercise located in [Here](https://artemis-test2.artemis.in.tum.de/course-management/529/programming-exercises/2033)
3. Open the `online editor` and submit the `assignment repository` (this won't complete and report about 
![image](https://user-images.githubusercontent.com/45761921/198254227-84c64fc5-c028-45a9-bea8-fcd56c8c8437.png)
)
4. Navigate to the related Build Plan [Here](https://jenkins-test.artemis.in.tum.de/job/SE101TESTUNICODEREMOVAL/job/SE101TESTUNICODEREMOVAL-ARTEMISTESTUSER11/configure)
5. Add the line `sed -i 's/[^[:print:]]//g' test-reports/tests-results.xml` to the pipeline at line 71 and save
6. Resubmit the `assignment repository` (this should now complete and report the results)

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|


### Screenshots
